src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -234,27 +287,35 @@ private void ClearProjectErrors(Solution solution, ProjectId projectId)
             // remove all document errors
             foreach (var documentId in project.DocumentIds)
             {
-                ClearDocumentErrors(solution, projectId, documentId);
+                ClearBuildOnlyDocumentErrors(solution, projectId, documentId);
             }
         }
 
-        private void ClearDocumentErrors(Solution solution, ProjectId projectId, DocumentId documentId)
+        private void ClearBuildOnlyDocumentErrors(Solution solution, ProjectId? projectId, DocumentId? documentId)
             => RaiseDiagnosticsRemoved(documentId, solution, projectId, documentId);
 
         public void AddNewErrors(ProjectId projectId, DiagnosticData diagnostic)
         {
             // capture state that will be processed in background thread.
             var state = GetOrCreateInProgressState();
 
-            _taskQueue.ScheduleTask("Project New Errors", () => state.AddError(projectId, diagnostic), CancellationToken.None);
+            _taskQueue.ScheduleTask("Project New Errors", async () =>
+            {
+                await ReportPreviousProjectErrorsIfRequiredAsync(projectId, state).ConfigureAwait(false);
+                state.AddError(projectId, diagnostic);
+            }, CancellationToken.None);
         }
 
         public void AddNewErrors(DocumentId documentId, DiagnosticData diagnostic)
         {
             // capture state that will be processed in background thread.
             var state = GetOrCreateInProgressState();
 
-            _taskQueue.ScheduleTask("Document New Errors", () => state.AddError(documentId, diagnostic), CancellationToken.None);
+            _taskQueue.ScheduleTask("Document New Errors", async () =>
+            {
+                await ReportPreviousProjectErrorsIfRequiredAsync(documentId.ProjectId, state).ConfigureAwait(false);
+                state.AddError(documentId, diagnostic);
+            }, CancellationToken.None);
```

Comment:
aside, all these cancellationtoken.none's are unfortuante.  you should be able to use IThreadingCOntext (which you will have, or can easily import), and use .DisposalToken instead.  This has the *awesome* aspect that when the Roslyn pakcage is closed, all this work is canceled.

Reply:
Filed https://github.com/dotnet/roslyn/issues/47694 to do this in a follow-up change.